core/lib/Cardano/Address.hs

```diff
@@ -77,6 +76,10 @@ newtype Address = Address
     { unAddress :: ByteString
     } deriving stock (Generic, Show, Eq, Ord)
 instance NFData Address
+instance ToBech32 Address where
+    bech32 (Address k) = bech32With CIP5.addr k
```

Comment:
My thought was that it would be easiest to just not use a ToBech32 typeclass, and make a function instead.

But @KtorZ had a neat suggestion about using a typeclass for determining the CIP5 hrp. I quote him here. Choose what you think is best.

> - I like simple functions as well over type classes for encoding. I prefer a `addressToBech32` and `scriptToBech32` than a `ToBech32` type-class.
> - On the other hand I also like consistency, and consistency is easier to achieve with a type-class.
> - Beside consistency, another reason for using a type-class here would be to avoid repeating / duplicating CIP's prefixes everywhere. We could possibly get away with a single type class `class CIP5 t where hrpFor :: Proxy t -> HumanReadablePart` which gives you what prefix to use for a particular data-type.
> - Overall, a type-class to connect the CIP's prefix to a data-type makes sense to me here because we really are encoding a same capability of data-types that depends on the data-types themselves.

Reply:
I've tried a few things with the `CIP5` typeclass only, without a `ToBech32`. Also, for the `Address` type, the `CIP5` human-readable part also depends on the content, so a `Proxy` isn't enough there. But then the API goes weird, where you have to pass things twice. A bunch of functions instead of a `ToBech32` class would also be possible, but then you'd have about 5-6 of these, with the `AddressVerificationKeyHash` newtypes. It gets _really_ interesting with the `Shelley` newtype, as you'll need separate functions for `XPub` and `XPrv`. I think I rest my case that a `ToBech32` typeclass is the best abstraction level we'll get.